pkg/vmprovider/providers/vsphere/vmprovider_vm.go

```diff
@@ -215,7 +214,12 @@ func (vs *vSphereVMProvider) createVirtualMachine(
 
 	// BMV: This is about where we used to do this check but it prb make more sense
 	// to do earlier, as to limit wasted work.
-	allowed, createDeferFn := vs.vmCreateConcurrentAllowed(vmCtx)
+	maxDeployThreads, ok := vmCtx.Value(context.MaxDeployThreadsContextKey).(int)
+	if !ok {
+		return nil, fmt.Errorf("MaxDeployThreadsContextKey missing from context")
```

Comment:
I am wondering if it is okay to just add a warning and still allow the VM to be created if there's an error in parsing the key from the `Context`.  Blocking reconciles because we can't parse a non-essential field seems off.

On the other hand, I don't see a scenario where the context will not have the key.  I will leave this to your judgement.

Reply:
The reason for this is because, as you said, there should not be a case where it's missing. I honestly almost opted for a panic instead to ensure there's never a situation where it's missing. The logic has a default associated with it, so it should be fine. Plus, as BV mentioned, this will go away sooner than later once I or BV implements the task-based reconciles.

Reply:
Oh, and also, the other reason was how the value is calculated. I elected to do the calculation at the point of injection instead of at use. Thus at this point we don't want to be calling the `lib` package to get the config. If anything I'll switch this to a panic, but the point is it shouldn't be defaulting here, but where it's injected.

Reply:
This is also a side-effect of the provider model that tried to divorce the Kubernetes side of things from the backend when it's probably an unnecessary abstraciton. Once we reign that in a bit, we can start to move config back closer to implementation.

Reply:
As it is, I don't see a sane way to inject a default here that makes sense. As we both said, it should always be injected at this point, and I think we rely on that than defaulting to something that could easily mask if the actual, configured value _isn't_ being passed through correctly.